pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,574 @@
+		rm.repoLocker.Lock(pod.Namespace, false)
+		defer rm.repoLocker.Unlock(pod.Namespace, false)
+
+		volumeBackup := &arkv1api.PodVolumeBackup{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    backup.Namespace,
+				GenerateName: backup.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: arkv1api.SchemeGroupVersion.String(),
+						Kind:       "Backup",
+						Name:       backup.Name,
+						UID:        backup.UID,
+						Controller: boolptr.True(),
+					},
+				},
+				Labels: map[string]string{
+					arkv1api.BackupNameLabel: backup.Name,
+					arkv1api.BackupUIDLabel:  string(backup.UID),
+				},
+			},
+			Spec: arkv1api.PodVolumeBackupSpec{
+				Node: pod.Spec.NodeName,
+				Pod: corev1api.ObjectReference{
+					Kind:      "Pod",
+					Namespace: pod.Namespace,
+					Name:      pod.Name,
+					UID:       pod.UID,
+				},
+				Volume: volumeName,
+				Tags: map[string]string{
+					"backup":     backup.Name,
+					"backup-uid": string(backup.UID),
+					"pod":        pod.Name,
+					"pod-uid":    string(pod.UID),
+					"ns":         pod.Namespace,
+					"volume":     volumeName,
+				},
+				RepoPrefix: rm.repoPrefix,
+			},
+		}
+
+		// TODO should we return here, or continue with what we can?
+		if err := errorOnly(rm.arkClient.ArkV1().PodVolumeBackups(volumeBackup.Namespace).Create(volumeBackup)); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	var errs []error
+
+ForLoop:
+	for i := 0; i < len(volumesToBackup); i++ {
+		select {
+		case <-ctx.Done():
+			errs = append(errs, errors.New("timed out waiting for all PodVolumeBackups to complete"))
+			break ForLoop
+		case res := <-resultChan:
+			switch res.Status.Phase {
+			case arkv1api.PodVolumeBackupPhaseCompleted:
+				SetPodSnapshotAnnotation(pod, res.Spec.Volume, res.Status.SnapshotID)
+			case arkv1api.PodVolumeBackupPhaseFailed:
+				errs = append(errs, errors.Errorf("pod volume backup failed: %s", res.Status.Message))
+			}
+		}
+	}
+
+	return kerrs.NewAggregate(errs)
+}
+
+func (rm *repositoryManager) RestorePodVolumes(ctx context.Context, restore *arkv1api.Restore, pod *corev1api.Pod, log logrus.FieldLogger) error {
+	// get volumes to restore from pod's annotations
+	volumesToRestore := GetPodSnapshotAnnotations(pod)
+	if len(volumesToRestore) == 0 {
+		return nil
+	}
+
+	resultChan := make(chan *arkv1api.PodVolumeRestore)
+
+	informer := arkv1informers.NewFilteredPodVolumeRestoreInformer(
+		rm.arkClient,
+		restore.Namespace,
+		0,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(opts *metav1.ListOptions) {
+			opts.LabelSelector = fmt.Sprintf("%s=%s", arkv1api.RestoreNameLabel, restore.Name)
+		},
+	)
+	informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
+				pvr := obj.(*arkv1api.PodVolumeRestore)
+
+				if pvr.Status.Phase == arkv1api.PodVolumeRestorePhaseCompleted || pvr.Status.Phase == arkv1api.PodVolumeRestorePhaseFailed {
```

Comment:
Can this function be run in parallel for 2 pods for the same restore? We'll likely need to match PVRs for this restore's name/UID.